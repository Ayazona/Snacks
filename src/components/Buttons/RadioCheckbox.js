import React, { Component, PropTypes }        from 'react'
import Radium                                 from 'radium'
import svgRadioBtnBase                        from '../../assets/radioBtnBase.svg'
import svgRadioBtnChecked                     from '../../assets/radioBtnChecked.svg'
import svgCheckboxBase                        from '../../assets/checkboxBase.svg'
import svgCheckboxChecked                     from '../../assets/checkboxChecked.svg'

const BASE_STYLES = {
  input: {
    margin: '0 10px -7px 0',
    backgroundSize: 'contain',
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'top left',
    backgroundColor: 'transparent',
    appearance: 'none',
    '-moz-appearance': 'none',
    '-webkit-appearance': 'none',
    width: 22,
    height: 22,
  },
}

const STYLES = {
  radioBase: {
    ...BASE_STYLES.input,
    backgroundImage: `url(${svgRadioBtnBase.url})`,
    ':active': {
      backgroundImage: `url(${svgRadioBtnBase.url})`,
      border: '0 solid transparent'
    }
  },
  radioChecked: {
    backgroundImage: `url(${svgRadioBtnChecked.url})`,
    ':active': {
      backgroundImage: `url(${svgRadioBtnChecked.url})`,
      border: '0 solid transparent'
    }
  },
  checkboxBase: {
    ...BASE_STYLES.input,
    backgroundImage: `url(${svgCheckboxBase.url})`,
    ':active': {
      backgroundImage: `url(${svgCheckboxBase.url})`,
      border: '0 solid transparent'
    }
  },
  checkboxChecked: {
    backgroundImage: `url(${svgCheckboxChecked.url})`,
    ':active': {
      backgroundImage: `url(${svgCheckboxChecked.url})`,
      border: '0 solid transparent'
    }
  }
}

class RadioCheckbox extends Component {
  static propTypes = {
    aria          : PropTypes.shape({
      label         :PropTypes.string,
    }),
    btnType       : PropTypes.oneOf(['radio', 'checkbox']),
    id            : PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    isChecked     : PropTypes.bool,
    onClick       : PropTypes.func,
    styles        : PropTypes.shape({
      button           : PropTypes.object,
    }),
    value         : PropTypes.string,
  }

  static defaultProps = {
    isChecked   : false,
    btnType     : 'radio',
    styles      : {},
    aria        : {},
  }

  state = {
    isChecked   : this.props.isChecked,
    styles      : {}
  }

  componentDidMount() {
    this.onMount()
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.isChecked !== prevState.isChecked) {
      this.setState({styles: this.getStyles()})
    }

    if (this.props.isChecked !== prevProps.isChecked) {
      this.setState({isChecked: this.props.isChecked})
    }
  }

  onMount() {
    this.setState({styles: this.getStyles()})
  }

  getStyles() {
    const { btnType, styles } = this.props
    const { isChecked } = this.state
    const finalStyles = {}
    let isCheckedStyle = {}

    if (btnType === 'checkbox') {
      if (isChecked) { isCheckedStyle = STYLES.checkboxChecked }
      finalStyles.button = {...STYLES.checkboxBase, ...isCheckedStyle, ...styles.button}
    } else {
      if (isChecked) { isCheckedStyle = STYLES.radioChecked }
      finalStyles.button = {...STYLES.radioBase, ...isCheckedStyle, ...styles.button}
    }

    return finalStyles
  }

  onClick = (event) => {
    const { onClick } = this.props
    const { isChecked } = this.state

    this.setState({isChecked: !isChecked})
    if (onClick) { onClick(event, !isChecked) }
  }

  render() {
    const {
      aria,
      btnType,
      id,
      value,
    } = this.props

    return (
      <input
        id={id}
        type={btnType}
        onClick={this.onClick}
        style={this.state.styles.button}
        aria-label={aria.label}
        value={value}
      />
    )
  }
}

export default Radium(RadioCheckbox)
