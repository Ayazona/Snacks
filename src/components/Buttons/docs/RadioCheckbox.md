This component outputs a radio button or a checkbox.

The programming logic behind producing these buttons is the same. The difference is in the props `btnType`. It defaults to radio if undefined.

You can define whether the button starts in the pressed on unpressed state through the `isChecked` props.

### Label
Associating a label for every radio or checkbox button is very important for accessibility purposes. Use the `id` and `htmlFor` attributes to link them.

There are times when it makes sense to have screenreaders say something different to what the label displays. For those cases, pass an aria label using the aria shape defined in the props.

### Radio
The component defaults to a radio button is a `btnType` is undefined. You can also pass the value `radio`. Also, you can pass a callback through the props `onClick` that returns both the click event and the checked/unchecked state.

```js
function onClick(event, isChecked) {
  alert(`This button is ${isChecked ? '' : 'not '}checked`)
}

<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <RadioCheckbox id="radio1" onClick={onClick} />
    <label htmlFor="radio1">Click me with callback</label>
  </div>
  <div style={{display: 'flex'}}>
    <RadioCheckbox id="radio2" isChecked />
    <label htmlFor="radio2">Click me</label>
  </div>
</div>
```

### Checkbox
Pass a value of `checkbox` to `btnType` and a checkbox will be generated.

```js
function onClick(event, isChecked) {
  alert(`This button is ${isChecked ? '' : 'not '}checked`)
}

<div style={{display: 'flex', flexDirection: 'column'}}>
  <div style={{display: 'flex', marginBottom: 20}}>
    <RadioCheckbox id="checkbox1" btnType='checkbox' onClick={onClick} />
    <label htmlFor="checkbox1">Checkbox with callback</label>
  </div>
  <div style={{display: 'flex'}}>
    <RadioCheckbox id="checkbox2" btnType='checkbox' isChecked />
    <label htmlFor="checkbox2">Click me</label>
  </div>
</div>
```



