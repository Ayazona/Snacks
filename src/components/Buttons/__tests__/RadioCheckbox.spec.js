import RadioCheckbox from '../RadioCheckbox'
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'

describe('RadioCheckbox', () => {
  it('renders the correct button type', () => {
    const testCases = [
      { id: 1, btnType: 'checkbox' },
      { id: 2, btnType: 'radio' },
      { id: 3, btnType: undefined },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<RadioCheckbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it("renders the correct checked state", () => {
    const testCases = [
      { id: 1, isChecked: false },
      { id: 2, isChecked: true },
      { id: 3, isChecked: undefined },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<RadioCheckbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('incorporates user styles if passed', () => {
    const testCases = [
      { id: 1, styles: { button: { border: '2px solid purple' } } },
      { id: 2, styles: undefined },
    ]

    testCases.forEach(props => {
      const tree = renderer
        .create(<RadioCheckbox {...props} />)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
