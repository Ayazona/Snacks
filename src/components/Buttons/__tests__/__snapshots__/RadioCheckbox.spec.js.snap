// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioCheckbox incorporates user styles if passed 1`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={1}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "border": "2px solid purple",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox incorporates user styles if passed 2`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={2}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct button type 1`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={1}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="checkbox"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct button type 2`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={2}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct button type 3`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={3}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct checked state 1`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={1}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct checked state 2`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={2}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;

exports[`RadioCheckbox renders the correct checked state 3`] = `
<input
  aria-label={undefined}
  data-radium={true}
  id={3}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "-moz-appearance": "none",
      "-webkit-appearance": "none",
      "appearance": "none",
      "backgroundColor": "transparent",
      "backgroundImage": "url(undefined)",
      "backgroundPosition": "top left",
      "backgroundRepeat": "no-repeat",
      "backgroundSize": "contain",
      "height": 22,
      "margin": "0 10px -7px 0",
      "width": 22,
    }
  }
  type="radio"
  value={undefined}
/>
`;
